### PR TITLE
use compile_commands.json for clang linter

### PR DESCRIPTION
**Pull request by tex**
This parses compile_commands.json and uses specific compile command for clang. Works fine for me.

### CODE REVIEW

**Comment by tex** [comment]
also fixes #1163

**Comment by tex** [comment]
also fixes #1163

**Comment by tex** [comment]
I did not notice the #725 though.

**Comment by tex** [comment]
Also this works - but it is dependency on roxma/ncm-clang:
    autocmd BufEnter *.cpp,*.h,*.hpp,*.hxx let g:ale_cpp_clang_options = join(ncm_clang#compilation_info()['args'], ' ')

**Comment by esran** [comment]
I created #725 a while ago in order to use compile_commands with the various c/c++ linters. Unfortunately I've never had time to go back and finish it off properly. I've actually moved on; switched back to master and put the compile_commands stuff in my vimrc. Not sure when or if I'd be able to find the time to go back and finish stuff off (some edge cases and vader testing).

File: ale_linters/cpp/clang.vim [hunk 1]
```diff
@@ -9,15 +10,31 @@ function! ale_linters#cpp#clang#GetExecutable(buffer) abort
...
+        let l:paths = ale#c#FindLocalHeaderPaths(a:buffer)
+        " -iquote with the directory the file is in makes #include work for
+        "  headers in the same directory.
+        return ale#Escape(ale_linters#cpp#clang#GetExecutable(a:buffer))
+        \   . ' -S -x c++ -fsyntax-only '
+        \   . '-iquote ' . ale#Escape(fnamemodify(bufname(a:buffer), ':p:h')) . ' '
+        \   . ale#c#IncludeOptions(l:paths)
+        \   . ale#Var(a:buffer, 'cpp_clang_options') . ' -'
+    else
+        let l:c = readfile(l:build_dir . '/compile_commands.json')
```
Comment (w0rp) [inline ale_linters/cpp/clang.vim]: For implementing this, you ought to make the `readfile` part a command to execute, so the file can be read in the background. `cat` might work for Unix. For windows, I dunno, maybe `type`? The command should be part of a command chain.

Comment (tex) [reply ale_linters/cpp/clang.vim]: Why?

Comment (w0rp) [reply ale_linters/cpp/clang.vim]: `readfile` calls are blocking and you can read the file in the background instead, so it won't interrupt editing the file.

Comment (tex) [reply ale_linters/cpp/clang.vim]: Well, do you have something like that already used anywhere else so I can look how to do it?
Besides, readfile is native function, it should be pretty fast...This code gets executed only on some even (save, or something), does it makes any hit into user's perception?

**Comment by w0rp** [comment]
Closed due to inactivity.

File: ale_linters/cpp/clang.vim (hunk 1 above)
Comment (tbodt) [reply ale_linters/cpp/clang.vim]: `readfile` is very fast, that is true, but `json_decode` is not so fast.

File: ale_linters/cpp/clang.vim [hunk 2]
```diff
@@ -9,15 +10,31 @@ function! ale_linters#cpp#clang#GetExecutable(buffer) abort
...
+        let l:l = json_decode(l:c)
+        let l:b = fnamemodify(bufname(a:buffer), ':p')
+        for l:d in l:l
+            if l:d["file"] == l:b
```
Comment (tbodt) [inline ale_linters/cpp/clang.vim]: The filenames in the compile_commands.json files generated by ninja (and therefore meson) are relative paths, so l:d["file"] should be expanded to an absolute path as well to compare.

**Pull request closed without merging**